src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresGroupDAO.scala

```diff
@@ -0,0 +1,366 @@
+    samsql"update ${GroupTable.table} set ${g.updatedDate} = ${Instant.now()} where ${g.id} = (${workbenchGroupIdentityToGroupPK(groupId)})".update().apply()
+  }
+
+  def deleteGroup(groupName: WorkbenchGroupName)(implicit session: DBSession): Int = {
+    val g = GroupTable.syntax("g")
+
+    val maybeGroupPK = Try {
+      // foreign keys in accessInstructions and groupMember tables are set to cascade delete
+      // note: this will not remove this group from any parent groups and will throw a
+      // foreign key constraint violation error if group is still a member of any parent groups
+      samsql"delete from ${GroupTable as g} where ${g.name} = ${groupName} returning ${g.result.id}".map(_.get[GroupPK](g.resultName.id)).single().apply()
+    }.recoverWith {
+      case fkViolation: PSQLException if fkViolation.getSQLState == PSQLStateExtensions.FOREIGN_KEY_VIOLATION =>
+        Failure(new WorkbenchExceptionWithErrorReport(
+          ErrorReport(StatusCodes.Conflict, s"group ${groupName.value} cannot be deleted because it is a member of at least 1 other group")))
+    }.get
+
+    maybeGroupPK.foreach(removeAllMembersFromFlatGroup)
+    maybeGroupPK.size // this should be 0 or 1
+  }
+
+  private def memberClause(member: WorkbenchSubject, f: scalikejdbc.QuerySQLSyntaxProvider[scalikejdbc.SQLSyntaxSupport[GroupMemberFlatRecord], GroupMemberFlatRecord]): SQLSyntax = {
+    member match {
+      case subGroupId: WorkbenchGroupIdentity => samsqls"${f.memberGroupId} = (${workbenchGroupIdentityToGroupPK(subGroupId)})"
+      case WorkbenchUserId(userId) => samsqls"${f.memberUserId} = $userId"
+      case _ => throw new WorkbenchException(s"illegal member $member")
+    }
+  }
+
+  protected def queryForGroupPKs(members: Set[WorkbenchSubject])(implicit session: DBSession): List[GroupPK] = {
+
+    // group PK query
+    val memberGroupNames = members.collect {
+      case groupName: WorkbenchGroupName => groupName
+    }
+    val gpk = GroupTable.syntax("g")
+    val groupPKStatement = samsqls"""select ${gpk.id} as group_id from ${GroupTable as gpk} where ${gpk.name} in ($memberGroupNames)"""
+
+    // policy group PK query
+    val memberPolicyIdTuples = members.collect {
+      case policyId: FullyQualifiedPolicyId => samsqls"(${policyId.resource.resourceTypeName}, ${policyId.resource.resourceId}, ${policyId.accessPolicyName})"
+    }
+    val rt = ResourceTypeTable.syntax("rt")
+    val r = ResourceTable.syntax("r")
+    val p = PolicyTable.syntax("p")
+    val policyGroupPKStatement = samsqls"""select ${p.groupId} as group_id
+               from ${ResourceTypeTable as rt}
+               join ${ResourceTable as r} on ${rt.id} = ${r.resourceTypeId}
+               join ${PolicyTable as p} on ${r.id} = ${p.resourceId}
+               where (${rt.name}, ${r.name}, ${p.name}) in ($memberPolicyIdTuples)"""
+
+    // there are 4 scenarios: there are both member groups and policies, only one or the other or neither
```

Comment:
Optional nit: This feels explained in the match case below, so I don't know if we need the comment. (And could get out of date if we added a different type)

Reply:
I hear you but I am inclined to leave it because I feel like the `case (true, true)` stuff is messy and needs explaination